includes/connect/class-wc-stripe-connect.php

```diff
@@ -0,0 +1,185 @@
+<?php
+
+if ( ! defined( 'ABSPATH' ) ) {
+	exit;
+}
+
+if ( ! class_exists( 'WC_Stripe_Connect' ) ) {
+	/**
+	 * Stripe Connect class.
+	 */
+	class WC_Stripe_Connect {
+
+		const SETTINGS_OPTION = 'woocommerce_stripe_settings';
+
+		/**
+		 * Stripe connect api.
+		 *
+		 * @var object $api
+		 */
+		private $api;
+
+		/**
+		 * Constructor.
+		 *
+		 * @param WC_Stripe_Connect_API $api stripe connect api.
+		 */
+		public function __construct( WC_Stripe_Connect_API $api ) {
+			$this->api = $api;
+
+			add_action( 'admin_init', array( $this, 'maybe_handle_redirect' ) );
+		}
+
+		/**
+		 * Gets the OAuth URL for Stripe onboarding flow
+		 *
+		 * @param  string $return_url url to return to after oauth flow.
+		 *
+		 * @return string|WP_Error
+		 */
+		public function get_oauth_url( $return_url = '' ) {
+
+			if ( empty( $return_url ) ) {
+				$return_url = admin_url( 'admin.php?page=wc-settings&tab=checkout&section=stripe' );
+			}
+
+			if ( substr( $return_url, 0, 8 ) !== 'https://' ) {
+				return new WP_Error( 'invalid_url_protocol', __( 'Your site must be served over HTTPS in order to connect your Stripe account automatically.', 'woocommerce-gateway-stripe' ) );
+			}
+
+			$result = $this->api->get_stripe_oauth_init( $return_url );
+
+			if ( is_wp_error( $result ) ) {
+				return $result;
+			}
+
+			return $result->oauthUrl; // phpcs:ignore WordPress.NamingConventions.ValidVariableName.UsedPropertyNotSnakeCase
+		}
+
+		/**
+		 * Initiate OAuth connection request to Connect Server
+		 *
+		 * @param  bool $state Stripe onboarding state.
+		 * @param  int  $code  OAuth code.
+		 *
+		 * @return string|WP_Error
+		 */
+		public function connect_oauth( $state, $code ) {
+
+			$response = $this->api->get_stripe_oauth_keys( $code );
+
+			if ( is_wp_error( $response ) ) {
+				return $response;
+			}
+
+			return $this->save_stripe_keys( $response );
+		}
+
+		/**
+		 * Handle redirect back from oauth-init or credentials reset
+		 */
+		public function maybe_handle_redirect() {
+			if ( ! is_admin() ) {
+				return;
+			}
+
+			// redirect from oauth-init
+			if ( isset( $_GET['wcs_stripe_code'], $_GET['wcs_stripe_state'] ) ) {
+
+				$response = $this->connect_oauth( $_GET['wcs_stripe_state'], $_GET['wcs_stripe_code'] );
+				wp_safe_redirect( remove_query_arg( array( 'wcs_stripe_state', 'wcs_stripe_code' ) ) );
+				exit;
+			// redirect from credentials reset
+			} elseif ( isset( $_GET['reset_stripe_api_credentials'] ) ) {
```

Comment:
It isn't readily apparent to me this is protected by a nonce. Is it?

Reply:
Good catch! I missed this. I've added one here: https://github.com/woocommerce/woocommerce-gateway-stripe/pull/1267/commits/3e8b0e59183b9f70122a1c27c47d23f10849c1c0